Client/TabManagement/TabManagerImplementation.swift

```diff
@@ -99,6 +100,7 @@ class TabManagerImplementation: LegacyTabManager, Notifiable {
         defer {
             isRestoringTabs = false
             tabRestoreHasFinished = true
+            AppEventQueue.completed(.tabRestoration)
```

Comment:
I think you do still want this to be on the main thread, my only concern was the mechanism for achieving that was using GCD instead of async/await. 

How would you feel about flagging the EventQueue as a MainActor? This way the asserts can be removed and we can be confident all calls come in on the main thread?

Reply:
> flagging the EventQueue as a MainActor?

I think it could be a nice update to the class. I will probably need to dig into it a bit further though, there appear to be some implications that may have tradeoffs we'll want to consider in a codebase like this with such a significant mix of async + GCD code etc.

Testing it out initially for example turns up at least a couple things I think will need further discussion:

1. Not a big deal but we can't declare a shared/common instance as a global variable. So this: `let AppEventQueue = EventQueue<AppEvent>()` doesn't work. We could change to something like this:

```
@MainActor
struct AppEventQueue {
  static var `default` = EventQueue<AppEvent>()
}
```

but there is [a possible related bug in the Swift compiler](https://github.com/apple/swift/issues/58270) that could impact us if doing something like that. We also can't declare the shared instance as a simple static property on the main class due to the generics. 

We could probably just opt for a different approach for referencing the common instance, via `DependencyHelper

2. The callsites become a bit more cumbersome. Throughout the codebase we would be going from something like: `AppEventQueue.signal(event: .profileInitialized)`  to something like `Task { await AppEventQueue.default.signal(event: .profileInitialized) }`. There could be nuances to this that I don't

Reply:
> flagging the EventQueue as a MainActor?

I think it could be a nice update but there are a couple things I'll probably need to dig into a bit further. Initial testing turns up a few issues that may have significant tradeoffs:

The first one is pretty minor but we can't declare a shared global instance, so this: `let AppEventQueue = EventQueue<AppEvent>()` doesn't work and we would need to change to (perhaps) something like:

```
@MainActor
struct AppEventQueue {
  static var `default` = EventQueue<AppEvent>()
}
```

but there is [a potential Swift compiler bug](https://github.com/apple/swift/issues/58270) that could impact us if doing something like that. We also can't declare a shared instance as a static property on the class due to the generics.

We could instead opt for a different approach for referencing the shared instance, via `DependencyHelper` or dependency injection etc. but I worry that in practice, for something as ubiquitous (as this is intended to be), that will end up being a pain point.

More importantly, though, in a mixed codebase like this where there is so much code that hasn't yet been updated to use Swift Async, I believe that many of the callsites will become much more cumbersome. In many places I believe we'd be going from something like: `AppEventQueue.signal(event: .profileInitialized)` to `Task { await AppEventQueue.default.signal(event: .profileInitialized) }`.

And I believe that in some contexts things like
`guard AppEventQueue.activityIsCompleted(.tabRestoration) else { return }` won't even be possible (at least not without some async hoops).

So there are some implications that will probably need to be hashed out before we switch to Swift Async. I'm happy to investigate/discuss further though.

Reply:
I added a subtask to the Event Queue epic to explore Swift Async further, I think it might be a good potential topic for tech team discussion?